include/nbl/ext/DepthPyramidGenerator/DepthPyramidGenerator.h

```diff
@@ -0,0 +1,105 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_EXT_DEPTH_PYRAMID_GENERATOR_H_INCLUDED_
+#define _NBL_EXT_DEPTH_PYRAMID_GENERATOR_H_INCLUDED_
+
+#include "nabla.h"
+
+#include "nbl/video/IGPUImageView.h"
+#include "nbl/asset/format/EFormat.h"
+
+using namespace nbl;
+using namespace nbl::core;
+using namespace nbl::asset;
+using namespace nbl::video;
+
+namespace nbl
+{
+namespace ext
+{
+namespace DepthPyramidGenerator
+{
+
+// TODO: test: `E_MIPMAP_GENERATION_OPERATOR::BOTH`, `reundUpToPoTWithPadding = true`
+
+class DepthPyramidGenerator final
+{
+public:
+
+	enum class E_MIPMAP_GENERATION_OPERATOR
+	{
+		MAX, 
+		MIN,
+		BOTH // min goes to r, min to g
+	};
+	
+	enum class E_WORK_GROUP_SIZE
+	{
+		E16x16x1 = 16u,
+		E32x32x1 = 32u
+	};
+
+	struct Config
+	{
+		E_WORK_GROUP_SIZE workGroupSize = E_WORK_GROUP_SIZE::E32x32x1;
+		asset::E_FORMAT outputFormat = asset::E_FORMAT::EF_R32_SFLOAT;
+		E_MIPMAP_GENERATION_OPERATOR op = E_MIPMAP_GENERATION_OPERATOR::MAX;
+		uint32_t lvlLimit = 0u; //no limit when set to 0 (full mip chain)
+		bool roundUpToPoTWithPadding = false;
+	};
+
+	// inputDepthImageView - input texture
+	//outputDepthPyramidMips - array of created mipMaps
+	DepthPyramidGenerator(IVideoDriver* driver, IAssetManager* am, core::smart_refctd_ptr<IGPUImageView> inputDepthImageView,
+		core::smart_refctd_ptr<IGPUImageView>* outputDepthPyramidMips,
+		const Config& config = Config());
```

Comment:
keep particular workloads out of the constructor (one generator just creates the necessary shaders for a particular workgroup size and operator config

Reply:
if you need a utility for creating the image and imageviews for the depth pyramid, make it a static method